indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorStateManager.java

```diff
@@ -0,0 +1,386 @@
+
+  private boolean atLeastOneSuccessfulRun = false;
+  private boolean currentRunSuccessful = true;
+  private final CircularBuffer<TaskState> completedTaskHistory;
+  private final CircularBuffer<State> supervisorStateHistory; // From previous runs
+  private final ExceptionEventStore eventStore;
+
+  public SeekableStreamSupervisorStateManager(
+      State initialState,
+      SeekableStreamSupervisorConfig supervisorConfig
+  )
+  {
+    this.supervisorState = initialState;
+    this.healthinessThreshold = supervisorConfig.getHealthinessThreshold();
+    this.unhealthinessThreshold = supervisorConfig.getUnhealthinessThreshold();
+    this.healthinessTaskThreshold = supervisorConfig.getTaskHealthinessThreshold();
+    this.unhealthinessTaskThreshold = supervisorConfig.getTaskUnhealthinessThreshold();
+    Preconditions.checkArgument(
+        supervisorConfig.getMaxStoredExceptionEvents() >=
+        Math.max(healthinessThreshold, unhealthinessThreshold),
+        "numExceptionEventsToStore must be greater than or equal to both "
+        + "healthinessThreshold and unhealthinessThreshold"
+    );
+    this.eventStore = new ExceptionEventStore(
+        supervisorConfig.getMaxStoredExceptionEvents(),
+        supervisorConfig.isStoringStackTraces()
+    );
+    this.completedTaskHistory = new CircularBuffer<>(Math.max(healthinessTaskThreshold, unhealthinessTaskThreshold));
+    this.supervisorStateHistory = new CircularBuffer<>(Math.max(healthinessThreshold, unhealthinessThreshold));
+  }
+
+  /**
+   * Certain supervisor states can only be set if the supervisor hasn't had a successful iteration yet.  This function
+   * checks if there's been at least one successful iteration if needed and sets supervisor state to an appropriate
+   * new state.
+   */
+  public Optional<State> setState(State newState)
+  {
+    if (newState.isOnlySetWhenNoSuccessfulRunYet()) {
+      if (!atLeastOneSuccessfulRun) {
+        supervisorState = newState;
+        return Optional.of(newState);
+      } else {
+        return Optional.absent();
+      }
+    } else if (newState.equals(State.SUSPENDED)) {
+      atLeastOneSuccessfulRun = false; // We want the startup states again after being suspended
+    }
+    this.supervisorState = newState;
+    return Optional.of(newState);
+  }
+
+  public void storeThrowableEvent(Throwable t)
+  {
+    if (t instanceof PossiblyTransientStreamException && atLeastOneSuccessfulRun) {
+      t = new TransientStreamException(t);
+    }
+
+    eventStore.storeThrowable(t);
+    currentRunSuccessful = false;
+  }
+
+  public void storeCompletedTaskState(TaskState state)
+  {
+    completedTaskHistory.add(state);
+  }
+
+  public void markRunFinishedAndEvaluateHealth()
+  {
+    if (currentRunSuccessful) {
+      atLeastOneSuccessfulRun = true;
+    }
+
+    State currentRunState = State.RUNNING;
+
+    for (Map.Entry<Class, Queue<ExceptionEvent>> events : eventStore.getNonTransientRecentEvents().entrySet()) {
```

Comment:
I don't think this is clearing error conditions properly. Let's say that you have `maxStoredExceptionEvents` set to something very high, such that `ExceptionEventStore.storeThrowable()` never removes any items from `recentEventsQueue` or `recentEventsMap`. Then let's say you have `unhealthinessThreshold` set to 3 and you accumulate 3 `TransientStreamException` events - It'll drop into a LOST_CONTACT_WITH_STREAM state. Then next run, say there were no errors encountered on the run - it'll go back into a RUNNING state. But the next run if you get another TransientStreamException, it'll immediately drop back into a LOST_CONTACT_WITH_STREAM state. Again assuming `maxStoredExceptionEvents` is very high and is never hit, you could have a stream running successfully for hours, and then on the next TransientStreamException it goes into an unhealthy state without respecting `unhealthinessThreshold`. Does this behavior sound like what would happen?